File: RT-SOC.py
# usage: RT-SOC.py [-h] p d mp s sp

# Real Time Catan, CMD Line Version

# positional arguments:
#   p           The period of time, in seconds, a turn will last at the beginning of the game
#   d           The amount the period of time, in seconds, a turn decreases by per turn
#   mp          The minimum amount of time a turn can last
#   s           Seasons are enabled
#   sp          The number of turns a season lasts

# optional arguments:
#   -h, --help  show this help message and exit

import dice
import argparse
import time
import os
import pprint as pp

# Global Variables
SEASONS_NAMES = ['Summer, +1 Wheat', 'Fall, +1 Timber', 'Winter, +1 Rock', 'Spring, +1 Sheep']
NUMBER_OF_SEASONS = len(SEASONS_NAMES)

# Main Game Loop
def main():
    # Parse command line arguments
    parser = argparse.ArgumentParser(description='Real Time Catan, CMD Line Version')
    parser.add_argument('turnPeriod', metavar='p', type=float, nargs=1, 
        help='The period of time, in seconds, a turn will last at the beginning of the game')
    parser.add_argument('turnPerDecay', metavar='d', type=float, nargs=1,
        help='The amount the period of time, in seconds, a turn decreases by per turn')
    parser.add_argument('minTurnPeriod', metavar='mp', type=float, nargs=1,
        help='The minimum amount of time a turn can last')
    parser.add_argument('seasonsOn', metavar='s', type=bool, nargs=1,
        help='Seasons are enabled')
    parser.add_argument('seasonPeriod', metavar='sp', type=int, nargs=1,
        help='The number of turns a season lasts')

    args = parser.parse_args()

    # Initialize the state machine and the display object
    sm = state()
    disp = cmdDisplay()

    sm.turnPeriod = args.turnPeriod[0]
    sm.turnPerDecay = args.turnPerDecay[0]
    sm.minTurnPeriod = args.minTurnPeriod[0]
    sm.seasonsOn = args.seasonsOn[0]
    sm.seasonPeriod = args.seasonPeriod[0]

    sm.start()

    # Update the state machine, and display on a loop
    while(True):
        sm.update()
        disp.update(sm)

class state:
    def __init__(self):
        self.turnNumber = 0
        self.turnPeriod = 100.0
        self.turnPerDecay = 1.0
        self.minTurnPeriod = 10

        self.seasonsOn = True
        self.seasonPeriod = 10

        self.diceRoll = self._rollDice()
        self.turnBegan = 0.0
        self.turnEnds = 0.0
        self.season = 0 # Summer, Fall, Winter, Spring
    
    # start(self)
    # Starts the game
    def start(self):
        self.turnNumber = 0
        self.turnEnds = time.time()
        self.turnPeriod = self.turnPeriod + self.turnPerDecay
        self._nextTurn()

    # update(self)
    # Checks if the state machine needs to be updated, then updates it
    def update(self):
        currTime = time.time()
        if (currTime - self.turnBegan) > self.turnPeriod:
            self._nextTurn()

    def getTimeLeftInTurn(self):
        return self.turnEnds - time.time()

    # _nextTurn(self)
    # Updates the state machine to reflect the values of the next turn
    def _nextTurn(self):
        self.turnNumber = self.turnNumber + 1
        self.season = int(self.turnNumber / self.seasonPeriod) % NUMBER_OF_SEASONS

        self.turnPeriod = self.turnPeriod - self.turnPerDecay
        if (self.turnPeriod < self.minTurnPeriod):
            self.turnPeriod = self.minTurnPeriod

        self.turnBegan = self.turnEnds
        self.turnEnds = self.turnBegan + self.turnPeriod
        
        self.diceRoll = self._rollDice()

    def _rollDice(self):
        return dice.roll('2d6')


class cmdDisplay:
    def __init__(self):
        pass

    def update(self, sm):
        os.system('cls')
        print("Dice Roll is " + str(sm.diceRoll) + " = " + str(sum(sm.diceRoll)))
        print("Turn #: " + str(sm.turnNumber))
        print("Turn Period: " + str(sm.turnPeriod))
        if sm.seasonsOn:
            print("Season: " + SEASONS_NAMES[sm.season])

        print("")
        print("Time Left: " + str(sm.getTimeLeftInTurn()))

        time.sleep(0.005)
    
if __name__ == '__main__':
    #start_time = time.time()
    main()
    #print("%s seconds" % (time.time() - start_time))

File: RT-SOC-F/RT-SOC.f90
! usage: ./RT-SOC
! Real Time Catan, CMD Line Version ~FORTRAN Edition~

PROGRAM RT_SOC
IMPLICIT NONE
CHARACTER(20) :: season_names(4)
CHARACTER(1) :: tmp
INTEGER :: number_of_seasons,p,d,mp,roll,turn,season
LOGICAL :: season_on,season_unassigned=.TRUE.
REAL :: harvest

season_names = ['Summer, +1 Wheat','Fall, +1 Timber ','Winter, +1 Stone','Spring, +1 Sheep']
number_of_seasons = size(season_names)

WRITE(*,'(A)') '     ___                                      ________         '
WRITE(*,'(A)') '  __(   )__                ___            ___(        )___     '
WRITE(*,'(A)') ' (_________)              {___|          (___          ___)    '
WRITE(*,'(A)') '                         _____|_____         (________)        '
WRITE(*,'(A)') '       Real Time        )__   |   __)                          '
WRITE(*,'(A)') '   Settlers of Catan    )__   |   __)                          '
WRITE(*,'(A)') '   ~FORTRAN Edition~    )__   |   __)                          '
WRITE(*,'(A)') '                        )_____|_____)                          '
WRITE(*,'(A)') '                     |________|_________/|                     '
WRITE(*,'(A)') '_____________________|__________________/______________________'
WRITE(*,'(A)') '                   _________              ___                  '
WRITE(*,'(A)') '      ____                      ______              ______     '
WRITE(*,'(A)') '                                                               '
WRITE(*,'(A)') 'Enter the initial period of time (seconds) a turn will last at the beginning of the game'
READ(*,*) p
WRITE(*,'(A)') 'Enter the amount of time (seconds) the turn time will decrease per turn'
READ(*,*) d
WRITE(*,'(A)') 'The minimum amount of time a turn can last'
READ(*,*) mp
WRITE(*,'(A)') 'Enable seasons? Y/N'
READ(*,*) tmp
DO WHILE (season_unassigned)
	IF ((tmp.EQ.('Y')).OR.(tmp.EQ.('y'))) THEN
		season_on=.TRUE.
		season_unassigned=.FALSE.
	ELSE IF ((tmp.EQ.('N')).OR.(tmp.EQ.('n'))) THEN
		season_on=.FALSE.
		season_unassigned=.FALSE.
	ELSE
		WRITE(*,'(A)') 'Invalid input - Enable seasons? Y/N'
		READ(*,*) tmp
		CYCLE
	END IF
END DO
CALL RANDOM_SEED()
CALL RANDOM_NUMBER(harvest)
season=NINT(harvest*4)
turn=0
WRITE(*,'(A)') 'Game is starting in'
WRITE(*,'(A)') '5...'
CALL SLEEP(1)
WRITE(*,'(A)') '4...'
CALL SLEEP(1)
WRITE(*,'(A)') '3...'
CALL SLEEP(1)
WRITE(*,'(A)') '2...'
CALL SLEEP(1)
WRITE(*,'(A)') '1...'
CALL SLEEP(1)
DO WHILE(.NOT.season_unassigned)
	turn=turn+1
	CALL dice_roll(roll)
	WRITE(*,'(A)') '_____________________________________'
	WRITE(*,'(A,i5)') 'Turn: ',turn
	IF (season_on) WRITE(*,'(A)') season_names(season)
	WRITE(*,'(A,i5)') 'Roll: ',roll
	IF ((p-d*(turn-1)).GT.mp) THEN
		WRITE(*,'(A,i5)') 'Turn Duration: ',p-d*(turn-1)
		WRITE(*,'(A)') '_____________________________________'
		CALL SLEEP(p-d*(turn-1))
	ELSE
		WRITE(*,'(A,i5)') 'Turn Duration: ',mp
		WRITE(*,'(A)') '_____________________________________'
		CALL SLEEP(mp)
	END IF

	season=season+1
	IF (season.GT.number_of_seasons) season=1
END DO

END PROGRAM 

SUBROUTINE dice_roll(roll)
	INTEGER :: minimum,maximum,roll1,roll2,roll
	REAL :: die1,die2
	CALL RANDOM_NUMBER(die1)
	CALL RANDOM_NUMBER(die2)
	roll = NINT(die1*6+0.5)+NINT(die2*6+0.5)
END SUBROUTINE
